src/ast.c

```diff
@@ -1277,6 +1161,85 @@ JL_DLLEXPORT jl_value_t *jl_expand_stmt(jl_value_t *expr, jl_module_t *inmodule)
     return jl_expand_stmt_with_loc(expr, inmodule, "none", 0);
 }
 
+
+//------------------------------------------------------------------------------
+// Parsing API and utils for calling parser from runtime
+
+// Internal C entry point to parser
+// `text` is passed as a pointer to allow raw non-String buffers to be used
+// without copying.
+JL_DLLEXPORT jl_value_t *jl_parse(const char *text, size_t text_len, jl_value_t *filename,
+                                  size_t offset, jl_value_t *options)
+{
+    static jl_value_t *parse_func = NULL;
+    if (!parse_func && jl_core_module) {
+        jl_value_t *compiler = jl_get_global(jl_core_module, jl_symbol("Compiler"));
+        if (compiler && jl_is_module(compiler)) {
+            parse_func = jl_get_global((jl_module_t*)compiler, jl_symbol("parse"));
+        }
+    }
+    if (!parse_func) {
+        // Directly call the builtin parser during bootstrap.
+        jl_value_t *result = jl_fl_parse(text, text_len, filename, offset, options);
+        return result;
+    }
+    jl_value_t **args;
+    JL_GC_PUSHARGS(args, 5);
+    args[0] = parse_func;
+    args[1] = (jl_value_t*)jl_alloc_svec(2);
+    jl_svecset(args[1], 0, jl_box_uint8pointer((uint8_t*)text));
+    jl_svecset(args[1], 1, jl_box_ulong(text_len));
+    args[2] = filename;
+    args[3] = jl_box_ulong(offset);
+    args[4] = options;
+    jl_ptls_t ptls = jl_get_ptls_states();
+    size_t last_age = ptls->world_age;
+    ptls->world_age = jl_world_counter;
+    jl_value_t *result = jl_apply(args, 5);
+    ptls->world_age = last_age;
+    args[0] = result; // root during error checks below
+    if (!jl_is_svec(result)) {
+        jl_type_error("jl_parse", (jl_value_t*)jl_simplevector_type, result);
+    }
+    else if (jl_svec_len(result) != 2 || !jl_is_expr(jl_svecref(result, 0)) ||
+             !jl_is_long(jl_svecref(result, 1))) {
+        jl_error("Result from parser should be `svec(a::Expr, b::Int)`");
+    }
```

Comment:
Somewhat more canonical form:
```suggestion
JL_TYPECHK(parse, simplevector, result);
if (jl_svec_len(result) != 2)
    jl_error("Result from parser should be `svec(a::Expr, b::Int)`");
JL_TYPECHK(parse, expr, jl_svecref(result, 0));
JL_TYPECHK(parse, long, jl_svecref(result, 1));
```

Reply:
Oh right, thanks.

I tried something like this but `JL_TYPECHK(parse, svec, result)` didn't work, and I didn't realize that `simplevector` could be substituted.